### ISSUE TITLE
[OJS] Assign participant form list issue

### ISSUE THREAD

**Issue by vasylOstrovskyi** [pkp/pkp-lib#4900]
I use OJS 3.2.1-1 with three languages enabled. The Assign participant form lists incorrect number of users in selected user group.

**To Reproduce**
1. Enable multiple languages (in my case, 3 languages are active).
2. Choose any submission and click "assign"
3. Select some user group (e.g. editors) and click "Search"
4. I have 24 members of that group, but instead ot 20 I see only 4 persons in the list; however, in the status bar, I see "20 of 216". Clicking "Load more" I get 4 more items and "40 of 216" etc.
5. Sorting of items does not correspond to the current locale and is in general wrong.

**Additional information**
I examined the code and found that the issue is related to the function filterUsersNotAssignedToStageInUserGroup in lib/pkp/classes/user/UserStageAssignmentDAO.inc.php
I changed the sql statement in the line 78 from "SELECT u.* ..." to SELECT DISTINCT u.* ..." and this fixed the problem for me, but I do not think that this is the correct solution since the sorting is still unsatisfactory.

**Comment by asmecher** [comment]
@vasylOstrovskyi, is that version 3.1.2-1?

**Comment by janiosarmento** [comment]
I think this issue still exists on 3.3.

<img width="802" alt="1" src="https://user-images.githubusercontent.com/7367281/127914250-db968183-f4c4-40da-a0ca-cbcce0cf876d.png">

<img width="807" alt="2" src="https://user-images.githubusercontent.com/7367281/127914272-c92614b1-d1f1-40b5-869b-f4ef915c41d9.png">

<img width="977" alt="version-history" src="https://user-images.githubusercontent.com/7367281/127914452-d980e45b-7582-4d2e-a22e-e221af6485af.png">

**Comment by NateWr** [comment]
@asmecher maybe this will be fixed by your work on the user entity dao?

**Comment by asmecher** [comment]
Yes, it should be -- that whole toolset is being rewritten. I'll assign this to 3.4 and leave it open for a little experimentation/testing to make sure.

**Comment by pmangahis** [comment]
Kinda related to the above and hopefully, this is going to be addressed in the rewrite, but a client has pointed out that number of users that show up in the results is inconsistent. For some roles, only 2 users would appear and they'd have to click load more multiple times to have the users appear. With a larger editorial team, this would involve a lot of unnecessary clicking.

**Comment by asmecher** [comment]
@pmangahis, if you have a test condition for this, please hang onto it and we'll come back to it for verification before 3.4 is released.

**Comment by asmecher** [comment]
[sorry, hit the wrong button there]

**Comment by retostauffer** [comment]
I was facing the same issue in OJS `3.3.0-7` (migrated from OJS 2); we have `7` users but find `116` leading to the problem explained above (Assign Participants seems to be broken). I have uploaded a test-script and the output as a gist:

* Check <https://gist.github.com/retostauffer/c8b662e9216b8e479142216acf48b596>
* `test_query.sh` shows the queries, `test_query_result.txt` the corresponding output (live system).

## What causes the problem

The query which gets the users does a series of `LEFT JOIN`s (see gist) to join information from the tables `users`, `user_user_groups`, `stage_assignments`, `user_group_stage`, and `user_settings` such as "given name" and "family name". For some users we have _multiple definitions_ of specific settings in `user_settings` (such as "given name" and "family name").

Thus the `LEFT JOIN` creates multiple rows for each of the 7 users. As shown on [gist#file-test_query_result-txt](https://gist.github.com/retostauffer/c8b662e9216b8e479142216acf48b596#file-test_query_result-txt) this results in `116` rows (`SQL1` query).

I modified the query to count how often each user occurs (see [gist#file-test_query_result-txt](https://gist.github.com/retostauffer/c8b662e9216b8e479142216acf48b596#file-test_query_result-txt)). 

**What the UX does**: The interface (user select grid) only shows each user once. However, if the first 20 entries (default list length) correspond to the very same user, the UX only shows one entry. When pressing "Load more" it runs the same query with "`... LIMIT 20 OFFSET 20`" but rather than loading 20 more users it loads only 0-20 depending on how many duplicates the query returns.

## Detailed insights

The last output on [gist#file-test_query_result-txt](https://gist.github.com/retostauffer/c8b662e9216b8e479142216acf48b596#file-test_query_result-txt) shows a detailed look into `user_settings` for one specific user where the problem can be seen.

## How I fixed it
 
As a hot fix I've cleared the `user_settings` table for specific users having the problem. However, not sure if this is a smart move?

```
DELETE FROM user_settings WHERE locale = 'en_US' AND NOT assoc_id IS NULL AND user_id IN (1, 10, 31, 59, 60, 61, 67, 72, 94330, 101055, 112812);
```

Maybe this analysis helps and it would be great to hear if my `DELETE` could be a bad move.

**Comment by NateWr** [comment]
Thanks @retostauffer! I think Alec's got this fixed for 3.4, but a hot-fix for 3.3 would be good

It sounds like your `DELETE` statement is removing the users' name information. If you only run one language in your nournal, `en_US`, then this may be old data and is safe to remove. But otherwise you're probably removing important data...

**Comment by retostauffer** [comment]
@NateWr thanks for your reply.

We have just run some additional tests. Seems that the data is restored once the user logs in; thus **my fix does not work** (or only temporary). I think it requires a fix on the `.php` side. If I can find time (not sure) I may have a look at the code to see if we can come up with a hotfix.

**Comment by NateWr** [comment]
Is the primary language of the site (not the journal) different from `en_US`? Every user needs a few required fields in that language as a fallback. This may be why the data is getting regenerated.

**Comment by retostauffer** [comment]
@NateWr that's not the problem, both `site` and `journal` run on the same locale (`en_US`). Reproducing the problem again (see [gist](https://gist.github.com/retostauffer/d6c0e32303e64b4e701f6056bb386999):

* First output of [this gist](https://gist.github.com/retostauffer/d6c0e32303e64b4e701f6056bb386999): after I've applied my `DELETE FROM ...` fix. Temporarly everything (UX) works as expected.
* Second output of [this gist](https://gist.github.com/retostauffer/d6c0e32303e64b4e701f6056bb386999): Same query after logging in again; new entries generated (due to login).

As shown it does not add an additional locale. But for each entry for `locale = "en_US"` it adds an additional entry with `assoc_type=  0` and `assoc_id = 0` (before I had/kept both `NULL`/`NULL`).

Maybe I should delete the `NULL`/`NULL`s? I just have no idea what these two fields are for.

**Comment by librariam** [comment]
Another feedback from PKP|PS related to Patricia's comment above - a journal is wondering whether there could be a control to set the number of names that show up, similar to "lists"? E.g. they have a stable number of 5 JE's and currently have to click "load more" twice for all to appear.

**Comment by AlejandraCasas** [comment]
I have the same comment Patricia had from another PKP|PS client.

**Comment by touhidurabir** [comment]
> The Assign participant form lists incorrect number of users in selected user group.

I believe the problem  got already resolved in `OJS-3.3.0-13` and in `main` branch as tested in both version with multiple locale installed(and enabled) but could not produce this issue . 

> For some roles, only 2 users would appear and they'd have to click load more multiple times to have the users appear. With a larger editorial team, this would involve a lot of unnecessary clicking.

Tested it with multiple(at least 5/6) Editors/Section Editors/Translators but it seems fine . Also Tested with more than 10 items and aloo items are showing in the list properly. It seems it is already fixed . 

@asmecher if no one requesting any more about this issue, it's good to be marked as resolved and completed .

**Comment by NateWr** [comment]
Thanks @touhidurabir!
